File: go/_alg/quick_sort.go
package main

import (
	"fmt"
	"math/rand"
)

func quickSort(nums []int) []int {
	if len(nums) < 2 {
		return nums
	}
	pos := partition(nums)
	quickSort(nums[:pos])
	quickSort(nums[pos+1:])
	return nums
}

func partitionRandom(nums []int, l, r int) int {
	nr := l + rand.Intn(r-l+1)
	nums[r], nums[nr] = nums[nr], nums[r]
	return partition(nums)
}

func partition(nums []int) int {
	l, r := 0, len(nums)-1
	pivot := nums[r]
	p := l - 1
	for i := l; i <= r; i++ {
		if nums[i] <= pivot {
			p++
			// swap
			nums[p], nums[i] = nums[i], nums[p]
		}
	}
	return p
}

func main() {
	nums := []int{5, 4, 3, 2, 1}
	//fmt.Println(partition(nums, 0, len(nums)-1))
	//fmt.Println(nums)
	quickSort(nums)
	fmt.Println(nums)
}

File: python/_alg/quick_sort.py
import random


def partition(arr, l, r):
    pivot = arr[r]
    i = l - 1
    for j in range(l, r+1):
        if arr[j] <= pivot:
            i += 1
            arr[i], arr[j] = arr[j], arr[i]
    return i


def partition_rand(arr, l, r):
    rand_pivot = random.randrange(l, r)
    arr[rand_pivot], arr[r] = arr[r], arr[rand_pivot]
    return partition(arr, l, r)


def quick_sort(arr, l, r):
    if l < r:
        pi = partition_rand(arr, l, r)
        quick_sort(arr, l, pi - 1)
        quick_sort(arr, pi + 1, r)


if __name__ == '__main__':
    arr = [5, 4, 3, 2, 1]
    quick_sort(arr, 0, 4)
    print(arr)
